README.md

```diff
@@ -45,58 +45,140 @@ This project can be utilized one of two ways, via docker or via a Makefile inclu
+work.
+
+* `make` v4.2 or later
+* `jq` v1.5 or later
+* `git` v2.20 or later
+* `bash` v4 or later
+* `sed` v4 or later
+* `awk` v4 or later
+* `grep` v3 or later
+* `xargs` v4.7 or later
+* `curl` v7 or later
+
+The Makefile is written with Debian-based systems in mind, including Ubuntu, and
+all those packages can be installed using `apt` from default repos. Other platforms
+ought to work fine, as long as these packages/versions can be installed. For example,
+MacOS is known to work, using `brew` to get a modern version of `bash` and GNU
+tools (there are many guides, [here is one](https://www.topbug.net/blog/2013/04/14/install-and-use-gnu-command-line-tools-in-mac-os-x/)).
+
+There are a couple make targets that call out to `apt` directly. You can avoid
+that requirement by pre-installing these packages using your own package manager,
+or simply do not invoke targets that rely on these packages:
+
+* `zip`
+* `nodejs`
+
+## Makefile "operating modes"
+
+When using the Makefile approach, the `include` directive is a shell command that
+retrieves the "bootstrap" Makefile from this project, saves it to the current working
+directory, and then echoes back the name of the saved file. Make reads the bootstrap
+Makefile and processes its directives and targets.
+
+The bootstrap Makefile is very barebones. Its primary purpose is to bootstrap this
+`tardigrade-ci` project so all the make targets are available. It does this through
+its own `include` directive. The bootstrap include directive processes a shell
+command with logic to determine whether to clone this project, and then echoes
+back the path to this project's Makefile.
+
+The bootstrap Makefile supports two "operating modes" that we're calling **user mode**
+and **developer mode**.
+
+In **user mode**, the bootstrap Makefile automatically clones the `tardigrade-ci`
+repo to a subdirectory of the calling project, e.g. `/{project}/tardigrade-ci/`.
+This feature is called **auto-init**. In this mode, the `tardigrade-ci` subdirectory
+is considered "owned" by the bootstrap Makefile. The bootstrap Makefile will
+manage and update the clone (through the `include` logic) to ensure the checkout
+matches the git branch or tag specified by the env `TARDIGRADE_CI_BRANCH`. By
+default, this env will resolve to either the version specified in the `Dockerfile`
+(if present) or the `master` branch (if the Dockerfile is missing or is not
+specifying a version). The user may override the default behavior by providing
+the env explicitly, e.g. `TARDIGRADE_CI_BRANCH=foo make help`.
+
+In **developer mode**, the user is responsible for cloning the `tardigrade-ci`
+repo to a sibling directory of the current project, e.g. `/{project}/` and
+`/tardigrade-ci/`. When the bootstrap Makefile detects this condition, it will
+include the tardigrade-ci Makefile from the user-managed clone. This makes it easier
+for a developer to edit the tardigrade-ci project, and test those changes from
+a calling project.
+
```

Comment:
> When the bootstrap Makefile detects this condition...

I'm unclear a bit as to how the detection happens in **developer mode**.  With `tardigrade-ci` being a sibling directory, would the simple Makefile of the `/{project}/` be different and reference the local `tardigrade-ci` clone?  Or do you use the same simple Makefile and the bootstrap logic handles it.  I may need to better understand the bootstrap logic changes...

Reply:
The detection logic/method is the same whether its developer or user mode. The bootstrap Makefile handles it, so the {project} Makefile does not change between developer/user mode. The flow is like this:

1. The {project} Makefile "includes" the bootstrap Makefile.
2. The bootstrap Makefile "includes" the tardigrade-ci Makefile, and has the logic to determine where the tardigrade-ci Makefile is. That location/path is what differentiates the user vs developer modes.

This flow is what I attempted to describe in the first two paragraphs of the new section "Makefile operating modes". Further suggestions welcome on how to clarify that language!

Reply:
This developer mode vs user mode is not really a new thing. It just wasn't documented before. So with or without the changes in this PR, these modes are the same.

What changes in this PR is that in **user mode**, previously, you had to run `make init` manually before running any of the tardigrade-ci Makefile targets. With this PR, that is automatic. The bootstrap determines whether it needs to run the init logic, and does so if needed, as part of the "include" directive.

There are no changes to how **developer mode** works in this PR, I don't think.

Reply:
Got it.  That explanation helped.